### PR TITLE
Feature/liayoo/state size mgmt

### PR DESCRIPTION
**Pull request by liayoo**
- Added 'state' rule property
  - A state rule consists of 'max_children' and 'ordering'.
  - If the # of children exceeds max_children, children will get evicted (deleted) according to the ordering (only 'FIFO' is supported for now)
  - State rules are only applied to the exactly matching paths 
  - As-is
```
{
  ".rule": {
    "write": true
  }
}
```
  - To-be
```
{
  ".rule": {
    "write": true,
    "state": {
      "max_children": 100,
      "ordering": "FIFO",
    }
  }
}
```
- Added state rules to native services

### CODE REVIEW

**Comment by platfowner** [comment]
Offline discussion summary:
- To extend state permission to max_height, max_size, max_bytes (from max_children) 
- To rename garbage collecting rule to 'gc_max_siblings' and change it's location from the parent state to the child state in the rule tree (/rules)
- Apply garbage collecting rule to the closest ancestor state rule only

**Comment by liayoo** [comment]
@platfowner As discussed, I have separated `gc_max_siblings` from `max_children` and used the former for the automatic garbage collection and the latter for the permission checking. Also, the hardcoded legacy garbage collection logic has been removed along with the `enableHardCodedGC` feature flag.
There will be a follow-up PR for extending the state rules by adding max_bytes, max_size, and max_height properties. 
PTAL & please let me know if you have any questions! Thanks!

File: common/constants.js [hunk 1]
```diff
@@ -465,10 +466,10 @@ function isNativeFunctionId(fid) {
  * @enum {string}
  */
 const ShardingProperties = {
-  LATEST: 'latest',
   LATEST_BLOCK_NUMBER: 'latest_block_number',
   PARENT_CHAIN_POC: 'parent_chain_poc',
   PROOF_HASH: 'proof_hash',
+  PROOF_HASH_GC_MAX_SIBLINGS: 'proof_hash_gc_max_siblings',
```
Comment (platfowner) [inline common/constants.js:472]: unused value?

File: common/constants.js [hunk 2]
```diff
@@ -385,6 +383,9 @@ const OwnerProperties = {
  */
 const RuleProperties = {
   WRITE: 'write',
+  STATE: 'state',
+  STATE_MAX_CHILDREN: 'max_children',
```
Comment (platfowner) [inline common/constants.js:387]: minor: what about just MAX_CHILDREN and GC_MAX_SIBLINGS?

File: common/path-util.js [hunk 3]
```diff
@@ -176,11 +175,11 @@ class PathUtil {
   }
 
   static getLatestShardReportPath(branchPath) {
-    return CommonUtil.appendPath(branchPath, ShardingProperties.LATEST);
+    return CommonUtil.appendPath(branchPath, ShardingProperties.LATEST_BLOCK_NUMBER);
   }
 
   static getLatestShardReportPathFromValuePath(valuePath) {
-    const branchPath = CommonUtil.formatPath(valuePath.slice(0, -2));
+    const branchPath = CommonUtil.formatPath(valuePath.slice(0, -3));
```
Comment (platfowner) [inline common/path-util.js:182]: +1

File: common/path-util.js [hunk 4]
```diff
@@ -176,11 +175,11 @@ class PathUtil {
   }
 
   static getLatestShardReportPath(branchPath) {
-    return CommonUtil.appendPath(branchPath, ShardingProperties.LATEST);
+    return CommonUtil.appendPath(branchPath, ShardingProperties.LATEST_BLOCK_NUMBER);
```
Comment (platfowner) [inline common/path-util.js:178]: +1

File: db/state-util.js [hunk 5]
```diff
@@ -427,6 +493,46 @@ function hasConfigLabelOnly(stateTreeObj, configLabel) {
   return true;
 }
 
+/**
+ * Returns a new rule tree created by applying the rule change to
+ * the current rule tree.
+ * @param {Object} curRuleTree current rule tree (to be modified by this rule)
+ * @param {Object} ruleChange rule change
+ */
+function applyRuleChange(curRuleTree, ruleChange) {
```
Comment (platfowner) [inline db/state-util.js:502]: nice adoption!

File: db/index.js [hunk 6]
```diff
@@ -715,6 +725,11 @@ class DB {
         return CommonUtil.returnTxResult(105, `Triggered function call failed`, 1, funcResults);
       }
     }
+    if (value !== null) {
+      // NOTE(liayoo): Only apply the state garbage collection rules when it's not a deletion.
+      const applyParentStateRuleRes = this.applyStateGarbageCollectionRule(ruleEvalRes.matched.state, localPath);
```
Comment (platfowner) [inline db/index.js:730]: s/applyParentStateRuleRes/applyStateGcRuleRes/?

File: db/index.js [hunk 7]
```diff
@@ -1722,68 +1758,133 @@ class DB {
...
+    if (stateNodeForReading.numChildren() <= gcMaxSiblings) {
+      return 0;
+    }
+    let numDeleted = 0;
+    const stateNodeForWriting = this.getRefForWriting(parentPath);
+    const childLabelList = stateNodeForWriting.getChildLabels();
+    const numChildren = childLabelList.length;
+    while (numChildren - numDeleted > gcMaxSiblings) {
+      const childLabel = childLabelList[numDeleted++];
+      stateNodeForWriting.setChild(childLabel, StateNode.fromJsObject(null));
```
Comment (platfowner) [inline db/index.js:1882]: Would it be possible to use writeDatabase() instead? This approach could have some advantages in general:
- Keep centralized use cases of setChild() for easier code maintenance (i.e., less bug-prone)
- More consistency in the underlying changes (e.g. no need to call updateStateInfoForAllRootPaths() explicitly)

File: db/index.js [hunk 8]
```diff
@@ -1431,29 +1457,39 @@ class DB {
...
-        matched.closestRule.config, matched.pathVars, data, newData, auth, timestamp);
-      if (!evalRuleRes) {
-        logger.debug(`[${LOG_HEADER}] evalRuleRes ${evalRuleRes}, ` +
-          `matched: ${JSON.stringify(matched, null, 2)}, data: ${JSON.stringify(data)}, ` +
-          `newData: ${JSON.stringify(newData)}, auth: ${JSON.stringify(auth)}, ` +
-          `timestamp: ${timestamp}\n`);
+      evalWriteRuleRes = !!this.evalWriteRuleConfig(
+        matchedWriteRules.closestRule.config, matchedWriteRules.pathVars, data, newData, auth, timestamp);
+      if (!evalWriteRuleRes) {
+        logger.error(`[${LOG_HEADER}] evalWriteRuleRes ${evalWriteRuleRes}, ` +
```
Comment (platfowner) [inline db/index.js:1475]: What bout logger.info() as this can happen by end-user's mistakes?

File: db/index.js [hunk 9]
```diff
@@ -1431,29 +1457,39 @@ class DB {
...
+            `matched: ${JSON.stringify(matchedWriteRules, null, 2)}, data: ${JSON.stringify(data)}, ` +
+            `newData: ${JSON.stringify(newData)}, auth: ${JSON.stringify(auth)}, ` +
+            `timestamp: ${timestamp}\n`);
+      }
+      evalStateRuleRes = this.evalStateRuleConfig(matchedStateRules.closestRule.config, newValue);
+      if (!evalStateRuleRes) {
+        logger.error(`[${LOG_HEADER}] evalStateRuleRes ${evalStateRuleRes}, ` +
```
Comment (platfowner) [inline db/index.js:1482]: ditto

File: db/index.js [hunk 10]
```diff
@@ -1221,7 +1238,16 @@ class DB {
...
-    this.writeDatabase([PredefinedDbPaths.VALUES_ROOT, ...CommonUtil.parsePath(receiptPath)], receipt);
+    const parsedPath = CommonUtil.parsePath(receiptPath);
+    this.writeDatabase([PredefinedDbPaths.VALUES_ROOT, ...parsedPath], receipt);
+    // Garbage collection for old receipts.
+    const matchedStateRule = this.matchRulePath(parsedPath, RuleProperties.STATE);
+    const closestRule = {
+      path: matchedStateRule.matchedRulePath.slice(0, matchedStateRule.closestConfigDepth),
+      config: getRuleConfig(matchedStateRule.closestConfigNode)
+    };
+    const applyParentStateRuleRes = this.applyStateGarbageCollectionRule({ closestRule }, parsedPath);
```
Comment (platfowner) [inline db/index.js:1249]: s/applyParentStateRuleRes/applyStateGcRuleRes/?

File: db/index.js [hunk 11]
```diff
@@ -1722,68 +1758,133 @@ class DB {
...
+  evalStateRuleConfig(stateRuleConfig, newValue) {
+    if (!CommonUtil.isDict(stateRuleConfig)) {
+      return true;
+    }
+    if (!CommonUtil.isDict(newValue)) {
+      return true;
+    }
+    const stateRuleObj = stateRuleConfig[RuleProperties.STATE];
+    if (CommonUtil.isEmpty(stateRuleObj)) {
+      return false;
```
Comment (platfowner) [inline db/index.js:1847]: is empty state rule not allowed?

File: db/index.js [hunk 12]
```diff
@@ -1722,68 +1758,133 @@ class DB {
...
+    if (!CommonUtil.isDict(stateRuleConfig)) {
+      return 0;
+    }
+    const stateRuleObj = stateRuleConfig[RuleProperties.STATE];
+    if (CommonUtil.isEmpty(stateRuleObj) || !stateRuleObj[RuleProperties.STATE_GC_MAX_SIBLINGS]) {
+      return 0;
+    }
+    const gcMaxSiblings = stateRuleObj[RuleProperties.STATE_GC_MAX_SIBLINGS];
+    // Check the number of children of the parent
+    const parentPathLen = matchedRules.closestRule.path.length - 1;
```
Comment (platfowner) [inline db/index.js:1867]: worth adding path.length === 0 case handling just in case?

File: integration/node.test.js [hunk 13]
```diff
@@ -349,18 +349,31 @@ describe('Blockchain Node', () => {
...
           },
-          "subtree_configs": []
+          "state": {
+            "matched_path": {
+              "target_path": "/apps/test/test_rule/some/path",
+              "ref_path": "/apps/test/test_rule/some/path",
+              "path_vars": {},
+            },
+            "matched_config": {
+              "config": null,
```
Comment (platfowner) [inline integration/node.test.js:373]: Could you add a test case with non-null config value?

File: integration/node.test.js [hunk 14]
```diff
@@ -614,18 +627,31 @@ describe('Blockchain Node', () => {
...
             },
-            "subtree_configs": []
+            "state": {
+              "matched_path": {
+                "target_path": "/apps/test/test_rule/some/path",
+                "ref_path": "/apps/test/test_rule/some/path",
+                "path_vars": {},
+              },
+              "matched_config": {
+                "config": null,
```
Comment (platfowner) [inline integration/node.test.js:651]: ditto

File: common/constants.js (hunk 1 above)
Comment (liayoo) [reply common/constants.js]: Oops, thanks! Garbage collected..

File: common/constants.js (hunk 2 above)
Comment (liayoo) [reply common/constants.js]: Done!

File: db/state-util.js (hunk 5 above)
Comment (liayoo) [reply db/state-util.js:502]: Thank you!!

File: db/index.js (hunk 6 above)
Comment (liayoo) [reply db/index.js]: Updated!

File: db/index.js (hunk 7 above)
Comment (liayoo) [reply db/index.js]: Ah, thanks for pointing that out. Using writeDatabase() sounds better. Updated!

File: db/index.js (hunk 8 above)
Comment (liayoo) [reply db/index.js]: Changed it to debug(). Thanks!

File: db/index.js (hunk 9 above)
Comment (liayoo) [reply db/index.js]: ditto

File: db/index.js (hunk 10 above)
Comment (liayoo) [reply db/index.js]: Done!

File: db/index.js (hunk 11 above)
Comment (liayoo) [reply db/index.js]: It was for a sanity check (at this point, stateRuleObj should be null or an object with a state property), but decided to return true just in case..

File: db/index.js (hunk 12 above)
Comment (liayoo) [reply db/index.js:1867]: Added a check for parentPathLen < 0. Thanks!

File: integration/node.test.js (hunk 13 above)
Comment (liayoo) [reply integration/node.test.js:392]: Done!!

File: integration/node.test.js (hunk 14 above)
Comment (liayoo) [reply integration/node.test.js:759]: ditto

**Comment by liayoo** [comment]
Addressed all comments, PTAL @platfowner

File: unittest/db.test.js [hunk 15]
```diff
@@ -1531,6 +1611,52 @@ describe("DB operations", () => {
         expect(node.db.setValue("/apps/test/shards/disabled_shard/path", 20).code).to.equal(0);
         expect(node.db.getValue("/apps/test/shards/disabled_shard/path")).to.equal(20)
       })
+
+      it("when writing more than gc_max_siblings in state rule config", () => {
```
Comment (platfowner) [inline unittest/db.test.js:1615]: nice test cases!

File: common/constants.js [hunk 16]
```diff
@@ -385,6 +383,9 @@ const OwnerProperties = {
  */
 const RuleProperties = {
```
Comment (platfowner) [inline common/constants.js:384]: minor: Could you leave a todo for other state info properties (max_height, max_size, max_bytes)?

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks for the nice implementations!

File: common/constants.js (hunk 16 above)
Comment (liayoo) [reply common/constants.js:384]: Right, done!

**Comment by liayoo** [comment]
Thanks for the helpful review & feedback. Now merging!
